src/algorithms/split_and_merge_algorithm.h

```diff
@@ -0,0 +1,149 @@
+#define BAYESMIX_ALGORITHMS_SPLIT_AND_MERGE_ALGORITHM_H_
+
+#include <Eigen/Dense>
+#include <memory>
+
+#include "algorithm_id.pb.h"
+#include "marginal_algorithm.h"
+#include "src/hierarchies/base_hierarchy.h"
+
+//! Template class for Split and Merge for conjugate hierarchies
+
+//! This class implements Split and Merge algorithm from Jain and Neal (2004)
+//! that generates a Markov chain on the clustering of the provided data.
+//!
+//! This algorithm requires the use of a `ConjugateHierarchy` object.
+class SplitAndMergeAlgorithm : public MarginalAlgorithm {
+ public:
+  // DESTRUCTOR AND CONSTRUCTORS
+  SplitAndMergeAlgorithm() = default;
+  ~SplitAndMergeAlgorithm() = default;
+
+  bool requires_conjugate_hierarchy() const override { return true; }
+
+  bayesmix::AlgorithmId get_id() const override {
+    return bayesmix::AlgorithmId::SplitMerge;
+  }
+
+  void read_params_from_proto(
+      const bayesmix::AlgorithmParams &params) override;
+
+ protected:
+  void print_startup_message() const override;
+
+  /* We need to update the parameters when we add or remove a datum from a
+   * cluster to esure that conditional_pred_lpdf returns a correct value.
+   */
+  bool update_hierarchy_params() override { return true; }
+
+  void sample_allocations() override;
+
+  void sample_unique_values() override;
+
+  Eigen::VectorXd lpdf_marginal_component(
+      std::shared_ptr<AbstractHierarchy> hier, const Eigen::MatrixXd &grid,
+      const Eigen::RowVectorXd &covariate) const override;
+
+  void compute_restricted_gs_data_idx(const unsigned int first_random_idx,
+                                      const unsigned int second_random_idx);
+
+  void compute_restricted_gs_unique_values(
+      const unsigned int first_random_idx,
+      const unsigned int second_random_idx);
+
+  /* Auxiliary function that computes the log of the ratio of the likelihoods
+   * and of the priors between the proposal and the original clustering.
+   * The two ratios are returned via the reference parameters.
+   * The parameter "split" represents if the function is called in the "split"
+   * case or in the "merge" case.
+   */
+  void compute_log_ratio_like_and_prior(const unsigned int first_random_idx,
+                                        const unsigned int second_random_idx,
+                                        bool split,
+                                        double &log_ratio_prior_prob,
+                                        double &log_ratio_likelihoods);
+
+  void split_or_merge(const unsigned int first_random_idx,
+                      const unsigned int second_random_idx);
+
+  /* Updates unique_values and allocations when a split or a merge proposal
+   * is accepted.
+   */
+  void proposal_update_allocations(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx,
+                                   const bool split);
+
+  void full_gibbs_sampling();
```

Comment:
Consider replacing `gibbs_sampling` with `gs` in both this name and the following one. You are already using the abbreviation everywhere else anyways. Regardless, I would make it explicit in the introductory text at the beginning of the file for clarity's sake: "...Gibbs Sampling (GS) ..."